lib/jsonapi/serializable/conditional_fields.rb

```diff
@@ -0,0 +1,65 @@
+module JSONAPI
+  module Serializable
+    module ConditionalFields
+      def self.prepended(klass)
+        klass.class_eval do
+          extend ClassMethods
+          prepend InstanceMethods
+          class << klass
+            attr_accessor :condition_blocks
+          end
+          self.condition_blocks = {}
+        end
+      end
+
+      module ClassMethods
+        def _register_condition(name, conditions)
+          condition_blocks[name.to_sym] =
+            if conditions.key?(:if)
+              conditions[:if]
+            elsif conditions.key?(:unless)
+              proc { !conditions[:unless].call }
```

Comment:
As discussed, it does work because instance variables are dynamically bound, and the proc/lambda will be `call`d in a context in which they are available.